mopidy-local: init at 3.1.1

ruuda opened the pull request:
###### Motivation for this change

[Mopidy-Local](https://github.com/mopidy/mopidy-local) is the successor to [Mopidy-Local-SQLite](https://github.com/mopidy/mopidy-local-sqlite) and
[Mopidy-Local-Images](https://github.com/mopidy/mopidy-local-images), which are both already packaged.

###### Things done

I had to make `gobject-introspection` a propagated build input, otherwise Mopidy imported by Mopidy-Local could not import `Gst`.

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Tested by building the following:
```nix
buildEnv {
  name = "mopidy-with-extensions-${mopidy.version}";
  paths = lib.closePropagation [
    pkgs.mopidy-iris
    pkgs.mopidy-local
  ];
  pathsToLink = [ "/${mopidyPackages.python.sitePackages}" ];
  buildInputs = [ makeWrapper ];
  postBuild = ''
    makeWrapper ${mopidy}/bin/mopidy $out/bin/mopidy \
      --prefix PYTHONPATH : $out/${mopidyPackages.python.sitePackages}
  '';
}
```
The `bin/mopidy` in there has a `mopidy local scan` command that successfully scanned my local collection.

ruuda committed:
mopidy-local: init at 3.1.1

Mopidy-Local is the successor to Mopidy-Local-SQLite and
Mopidy-Local-Images, which are already packaged. I had to make
gobject-introspection a propagated build input, otherwise
Mopidy-Local can't import Mopidy.

ruuda commented:
CC @rvolosatovs, who is the maintainer of the `mopidy-local-*` packages.

rvolosatovs reviewed (approved):
LGTM
Can we remove the other two as well then? I added them initially as a dependency to `mopidy-iris`, which does not need them anymore apparently.

ruuda committed:
mopidy-local-images: remove

This plugin has been merged into the newer "mopidy-local" plugin which I
just added. "mopidy-local-images" and "mopidy-local-sqlite" were added
originally for Mopidy Iris, but Iris now works with mopidy-local, and
does not need the older plugins any more.

ruuda commented:
> Can we remove the other two as well then? I added them initially as a dependency to mopidy-iris, which does not need them anymore apparently.

Mopidy-Iris does not need them, it works with the newer Mopidy-Local now. I added a commit to delete the older packages. The only thing I wonder, should this be recorded in a changelog somewhere? I can imagine people have these older plugins in their config, and that will break when upgrading.

mweinelt commented on pkgs/top-level/all-packages.nix:
```diff
@@ -21224,8 +21224,7 @@ in
     mopidy-gmusic
     mopidy-iris
-    mopidy-local-images
-    mopidy-local-sqlite
```
Let's add throw aliases to `aliases.nix` explaining the removal and migration steps.

```diff
diff --git a/pkgs/top-level/aliases.nix b/pkgs/top-level/aliases.nix
index 3e9cdc8e87e..634ecc3af0e 100644
--- a/pkgs/top-level/aliases.nix
+++ b/pkgs/top-level/aliases.nix
@@ -301,6 +301,8 @@ mapAliases ({
   mcgrid = throw "mcgrid has been removed from nixpkgs, as it's not compatible with rivet 3"; # added 2020-05-23
   mcomix = throw "mcomix has been removed from nixpkgs, as it's unmaintained"; # added 2019-12-10
   mirage = throw "mirage has been femoved from nixpkgs, as it's unmaintained"; # added 2019-12-10
+  mopidy-local-images = throw "mopidy-local-images has been removed as it's unmaintained. It's functionality has been merged into the mopidy-local extension."; # added 2020-10-18
+  mopidy-local-sqlite = throw "mopidy-local-sqlite has been removed as it's unmaintained. It's functionality has been merged into the mopidy-local extension."; # added 2020-10-18
   mysql-client = hiPrio mariadb.client;
   memtest86 = memtest86plus; # added 2019-05-08
   mesa_noglu = mesa; # added 2019-05-28
```

ruuda replied:
Ah, so this is how to deal with removing packages, thanks for pointing that out!

mweinelt commented on pkgs/applications/audio/mopidy/local.nix:
```diff
@@ -0,0 +1,30 @@
+{ stdenv
```
```suggestion
{ lib
```

mweinelt commented on pkgs/applications/audio/mopidy/local.nix:
```diff
@@ -0,0 +1,30 @@
+    python3Packages.pytest
+  ];
+
+  meta = with stdenv.lib; {
```
```suggestion
  meta = with lib; {
```

mweinelt commented on pkgs/applications/audio/mopidy/local.nix:
```diff
@@ -0,0 +1,30 @@
+    homepage = "https://github.com/mopidy/mopidy-local";
+    description = "Mopidy extension for playing music from your local music archive";
+    license = licenses.asl20;
+    maintainers = [ maintainers.ruuda ];
```
```suggestion
    maintainers = with maintainers; [ ruuda ];
```

mweinelt commented on pkgs/applications/audio/mopidy/mopidy.nix:
```diff
@@ -16,13 +16,25 @@ pythonPackages.buildPythonApplication rec {
-  ] ++ stdenv.lib.optional (!stdenv.isDarwin) dbus-python;
+  propagatedBuildInputs = [
+    gobject-introspection
+  ] ++ (
+    with pythonPackages; [
```
```suggestion
  ] ++ (with pythonPackages; [
```

mweinelt reviewed (changes requested):
Thanks for taking care of this. I've added a few remarks, nothing major.

We should convert the whole mopidy-shebang to python3 shortly, not necessarily in this pr.

ruuda commented:
Thanks for having a look @mweinelt. I addressed your comments and also rebased this PR on top of the latest master.

mweinelt commented:
Do you have experience on how these plugins interact with each other? I see we mix and match python2 and python3 there.

mweinelt commented:
https://github.com/NixOS/nixpkgs/pull/93400/commits/20b2e8eedca7d97dc9924693ee7c4a5fb83175fa

This commit should be split in two (local-images and local-sqlite) and use the format described in CONTRIBUTING.md, e.g.

```
mopidy-local-images: remove

The what and why here
```

ruuda committed:
mopidy-local-sqlite: remove

This plugin has been merged into the newer "mopidy-local" plugin which I
just added. "mopidy-local-images" and "mopidy-local-sqlite" were added
originally for Mopidy Iris, but Iris now works with mopidy-local, and
does not need the older ones any more.

ruuda commented:
> Do you have experience on how these plugins interact with each other? I see we mix and match python2 and python3 there.

I don’t think we do, the packages reference `pythonPackages`, but that is set here: https://github.com/NixOS/nixpkgs/blob/396f3407a28a0ce8ee537d6f4440d3e6cb359570/pkgs/applications/audio/mopidy/default.nix#L11

And the Python there is set to `python3` here: https://github.com/NixOS/nixpkgs/blob/396f3407a28a0ce8ee537d6f4440d3e6cb359570/pkgs/top-level/all-packages.nix#L22111-L22113

So it’s all running with Python 3.

> This commit should be split in two (local-images and local-sqlite)

Done.

mweinelt commented:
Result of `nixpkgs-review pr 93400` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>15 packages built:</summary>
  <ul>
    <li>mopidy</li>
    <li>mopidy-gmusic</li>
    <li>mopidy-iris</li>
    <li>mopidy-local</li>
    <li>mopidy-moped</li>
    <li>mopidy-mopify</li>
    <li>mopidy-mpd</li>
    <li>mopidy-mpris</li>
    <li>mopidy-musicbox-webclient</li>
    <li>mopidy-somafm</li>
    <li>mopidy-soundcloud</li>
    <li>mopidy-spotify</li>
    <li>mopidy-spotify-tunigo</li>
    <li>mopidy-tunein</li>
    <li>mopidy-youtube</li>
  </ul>
</details>

mweinelt commented:
As the `mopidy-local-{sqlite,images}` packages are broken on 20.09 I think this is worth backporting to `release-20.09`.
